Create rate limiter that interacts with Redis

earmenda opened the pull request:
# Purpose
Create a client side rate limiter which can share a state between fidesops instances using Redis.

# Changes
- Create new `RateLimiter` which can be configured with a period of `seconds` `minutes` `hours` `days`
- Implementation for a fixed window limiter that attempts to find an available bucket until timeout is reached

# Checklist
- [x] Update [`CHANGELOG.md`](https://github.com/ethyca/fidesops/blob/main/CHANGELOG.md) file
  - [x] Merge in main so the most recent `CHANGELOG.md` file is being appended to
  - [x] Add description within the `Unreleased` section in an appropriate category. Add a new category from the list at the top of the file if the needed one isn't already there.
  - [x] Add a link to this PR at the end of the description with the PR number as the text. example: [#1](https://github.com/ethyca/fidesops/pull/1)
- [x] Good unit test/integration test coverage


# Ticket

Fixes #1188

earmenda commented:
@galvana @adamsachs I moved this over from the fidesops repo and could use an initial review. open to any feedback

adamsachs commented on tests/ops/integration_tests/limiter/test_rate_limiter.py:
```diff
@@ -0,0 +1,46 @@
+    RateLimiterPeriod,
+    RateLimiterTimeoutException,
+)
+
```
these tests are nice. one thing i was thinking about that could be really cool to have coverage on is invoking the limiter from multiple threads, i.e. to evaluate that the shared cache is working as expected. i don't really know how feasible that is, so if it seems like it will be too hard, or if you don't think it makes sense, then that's fine

earmenda replied:
I think this is a good idea, it would be cool to test from multiple instances but this should be good enough. see `test_limiter_respecsts_rate_limit_multiple_threads`

adamsachs replied:
agreed, multiple instances is the _true_ test but given the constraints of our testing env, multiple threads seems plenty sufficient. thanks for adding this in!

adamsachs commented on src/fides/api/ops/service/connectors/limiter/rate_limiter.py:
```diff
@@ -0,0 +1,75 @@
+    Defines the periods which rate limiter supports for time buckets
+    """
+
+    SECONDS = ("seconds", 1)
```
nit: the plural's here confused me a bit. effectively this is "n requests per [second/minute/hour/day]", so it sort of seems to me like it should just be singular? but maybe i'm thinking about it incorrectly. not a big deal either way.

earmenda replied:
sure let's change it

adamsachs commented on src/fides/api/ops/service/connectors/limiter/rate_limiter.py:
```diff
@@ -0,0 +1,75 @@
+        start_time = time.time()
+        while time.time() - start_time < timeout_seconds:
+            current_seconds = int(time.time())
+            fixed_time_filter = int(current_seconds / period.factor) * period.factor
```
nice

earmenda commented:
> @earmenda this all looks pretty clean to me, i like the implementation.
> 
> i left a few minor comments, and i think you need to merge in main and re-run tests, but otherwise i think it's about ready to go.

@adamsachs thanks for taking a look. I addressed your comments! 

mypy failures being investigated here - https://github.com/ethyca/fides/pull/1226
